### PR TITLE
bpo-38490: statistics: Add covariance, Pearson's correlation, and simple linear regression

### PR DESCRIPTION
**Pull request by twolodzko**
This PR adds functions for calculating bivariate covariance and Pearson's correlation.

https://bugs.python.org/issue38490

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@twolodzko

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by twolodzko** [comment]
I've signed CLA & confirmed mail at least three times now, it does not seem to work.

**Comment by corona10** [comment]
I'd like to recommend you to open the PR after the discussion on https://bugs.python.org/issue38490 is finalized. :)

File: Doc/library/statistics.rst
Comment (taleinat) [inline Doc/library/statistics.rst]: Typo here.
```suggestion
:func:`linear_regression`  Intercept and slope for simple linear regression.
```

File: Doc/library/statistics.rst
Comment (taleinat) [inline Doc/library/statistics.rst]: This seems to be reversed: An exception would  be raised if the number of data points were different, right?

(The same goes fro the docs of the following function as well.)

**Review by taleinat — CHANGES_REQUESTED** [review]

**Comment by twolodzko** [comment]
I have made the requested changes; please review again

File: Doc/library/statistics.rst
Comment (taleinat) [inline Doc/library/statistics.rst]: This can be reworded to be shorter and simpler, in keeping with the style of this document. For example:

"Both inputs must be of the same length (no less than two), otherwise StatisticsError is raised."

The same goes for the other two functions.

File: Doc/library/statistics.rst
Comment (taleinat) [inline Doc/library/statistics.rst]: Use "Return the ..." rather than "Calculates", as elsewhere here.

File: Misc/NEWS.d/next/Library/2019-10-16-08-08-14.bpo-38490.QbDXEF.rst
Comment (taleinat) [inline Misc/NEWS.d/next/Library/2019-10-16-08-08-14.bpo-38490.QbDXEF.rst]: Please add a period at the end of the sentence.

Also, add "Patch by Tymoteusz Wołodźko" at the end! :)

**Review by taleinat — CHANGES_REQUESTED** [review]
Thanks for making the effort to add this and for making changes so quickly!

I've gone through the docs thoroughly and have a few requests (see inline comments.)

Additionally, in general, the use of "variables" in the suggested changes isn't currently present in these docs, and is a significant change in tone. Specifically, for many not coming from a Math or Statistics background, it will likely be confused with the common meaning of "variable" in programming, which is slightly but significantly different. Please reword to use "data" or "input data" as done elsewhere in these docs.

**Comment by taleinat** [comment]
Additionally, it seems prudent to state that the covariance computed is what is often called the "sample covariance". This is an estimate for the covariance of the two originating random variables. 

This would be similar to the existing docs for `mean()`: "Return the sample arithmetic mean of data."

File: Lib/statistics.py
Comment (taleinat) [inline Lib/statistics.py]: This algorithm unnecessarily computes the means and stddev of both data sets twice. It appears that one can simply use `slope = cov(regressor, dependent_variable) / (stdx * stdx)` instead! (This also means there's no need to compute `stdy` at all!)

Or slightly simpler:
```python
slope = cov(regressor, dependent_variable) / variance(regressor)
```

File: Lib/statistics.py
Comment (taleinat) [inline Lib/statistics.py:948]: This exception handling is unnecessary, since the call to `correlation()` will make the same check.

(If changing this as suggested above to use `covariance()` but not `correlation()`, only the final division operation would this exception handling.)

**Review by taleinat — CHANGES_REQUESTED** [review]
Please also add tests; the doc-tests are not enough.

See: `Lib/test/test_statistics.py`.

File: Doc/library/statistics.rst
Comment (taleinat) [inline Doc/library/statistics.rst]: Perhaps just "and not be constant" rather than "and stdev of both inputs needs to be greater than zero"? (Below for linear regression as well.)

File: Lib/statistics.py
Comment (taleinat) [inline Lib/statistics.py:96]: Maybe use the following to avoid decimal representation issues? Or use different values for the example so that "+ELLIPSIS" is enough.

```python3
>>> '%.2f, %.2f' % linear_regression(x, y)
'1.50, 0.10'
```

File: Lib/statistics.py
Comment (taleinat) [inline Lib/statistics.py]: Forgot to use "sample covariance" here too :)

File: Lib/test/test_statistics.py
Comment (taleinat) [inline Lib/test/test_statistics.py]: Copy/paste error here, testing `covariance()` twice but not `correlation()`. Below as well.

File: Lib/test/test_statistics.py
Comment (taleinat) [inline Lib/test/test_statistics.py]: Add test cases for covariance as well?

File: Doc/library/statistics.rst
Comment (taleinat) [inline Doc/library/statistics.rst]: ISTM this should also mention that this is for simple linear regression using the "ordinary least squares" method. (In the doc-string too.)

File: Doc/library/statistics.rst
Comment (taleinat) [inline Doc/library/statistics.rst]: This is hilarious! (Perhaps `use it to "predict" how many...`, since this isn't an actual prediction?)

**Review by taleinat — CHANGES_REQUESTED** [review]
Excellent! Just a few more small things.

**Comment by taleinat** [comment]
Also note that there's a whitespace issue in one of the files (that's causing the test failures here.) To find the issue quicly, you can run `make patchcheck`, or without make (e.g. on Windows), `python Tools/scripts/patchcheck.py`.

**Comment by twolodzko** [comment]
I have made the requested changes; please review again.

File: Lib/statistics.py
Comment (twolodzko) [reply Lib/statistics.py:96]: It seems to work now.

Comment (taleinat) [reply Lib/statistics.py:96]: Yes, but tests that sometimes fail can require a lot of effort down the road. We should make every effort to make tests robust.

**Comment by taleinat** [comment]
@rhettinger, I think this is now ready for your review.

**Review by taleinat — APPROVED** [review]
LGTM

Through possibly some doctests could use more robust handling of float representation issues, e.g. `0.1` vs. `0.099999999...`.

**Comment by twolodzko** [comment]
Is there anything I should do about the PR, or it just got lost into the void?

**Comment by taleinat** [comment]
@twolodzko, this PR isn't lost in the void. I regret the long delay.

This is awaiting review from @rhettinger and/or those from whom he requested a review, @tim-one and @stevendaprano.

File: Lib/statistics.py
Comment (rhettinger) [inline Lib/statistics.py]: Let's make these positional-only arguments.  We may well want to change the name of these parameters later.

File: Lib/statistics.py
Comment (rhettinger) [inline Lib/statistics.py]: I recommend using a named tuple here.

Comment (twolodzko) [reply Lib/statistics.py]: Done. I used `LinearRegression` namedtuple, if you have suggestion for better name, feel free to suggest something. I considered `RegressionParameters` or `LinearRegressionParameters`, but first is slightly ambiguous (there are different "regressions"), and second one rather lengthy. The only problem with `LinearRegression` is that it conflicts with scikit-learn class with same name, but I hardly see scenario where this would be a problem for anyone.

File: Lib/statistics.py
Comment (rhettinger) [inline Lib/statistics.py:913]: The parameter order seems backwards to me, here and through out the PR.  Traditionally, we say an equation for a line in is slope-intercept form as `y = mx + b` with the *m* preceding the *b*.

Comment (twolodzko) [reply Lib/statistics.py:913]: It depends on your background, [in statistics](https://en.wikipedia.org/wiki/Linear_regression#Introduction) the model is often written as y = β_0 + β_1 * x + ε , so the intercept is written before. Also, if now it is a named tuple, the order is of less importance. Of course, if you insist, I may change it.

**Comment by stevendaprano** [comment]
Thank you Tymoteusz for your work.

I agree with Raymond that using a namedtuple is a good idea here.

On Fri, Oct 09, 2020 at 12:52:16AM -0700, Raymond Hettinger wrote:
> The parameter order seems backwards to me, here and through out the 
> PR.  Traditionally, we say an equation for a line in is 
> slope-intercept form as `y = mx + b` with the *m* preceding the *b*.

This is true for most areas of mathematics except for statistics and 
finance. In those, the usual form of the linear regression equation is:

    y = a + bx

which as Raymond points out is the reverse of the usual form:

    y = mx + c

(or sometimes `mx + b` for the constant term).


See, for example:

https://www.statisticssolutions.com/conduct-interpret-linear-regression/

https://www.investopedia.com/terms/r/regression.asp



-- 
Steve

**Comment by stevendaprano** [comment]
On Fri, Oct 09, 2020 at 01:15:52AM -0700, Tymoteusz Wołodźko wrote:



> Also, if now it is a named tuple, the order is of 

> less importance. Of course, if you insist, I may change it.



I would prefer you keep the (intercept, gradient) order which is more 

conventional in statistics.

**Comment by twolodzko** [comment]
@stevendaprano @rhettinger should I apply any more changes, or is it ok as-is?

**Comment by taleinat** [comment]
The only current outstanding issue is the order of returned values from `linear_regression()`.

ISTM given the short discussion here that the order of `(intercept, gradient)` could go either way, and that in the context of statistics the current order seems fine. And since this will be a named tuple, the order is not as important as it would be otherwise. As long as the documentation and doc-string are consistent and clear on this, which they are, IMO this is fine.

I'd like to get @rhettinger's approval, or otherwise, for this before moving forward.

**Comment by stevendaprano** [comment]
Given that this is the statistics module, not linear algebra, I think we 
ought to stick to the common convention taught in stats classes, which 
is

    y = a + b x

i.e. (intercept, gradient) in that order. Unless Raymond strongly 
objects, I say go with this order.

Do we have time to add a new feature to this before feature-freeze? I 
would like to add two methods to the tuple returned:


    def predict_y(self, x):
        """Return the predicted y value for the given x.

        Returns the predicted response or dependent variable.
        """
        return self.intercept + self.gradient*x


    def predict_x(self, y):
        """Return the predicted x value for the given y.

        Returns the predicted explanatory or independent variable.
        """
        return (y - self.intercept)/self.gradient


Predicting the x or y values from the linear regression line are very 
common operations, I think it would be very useful to offer them without 
expecting users to create the function themselves.


-- 
Steve

**Comment by twolodzko** [comment]
@stevendaprano I'd be _against_ having `predict_x` as this is incorrect if this is a statistical model y = a + bx + noise, because it ignores the noise.

**Comment by twolodzko** [comment]
It's been two weeks and the discussion seems to got stuck. Is there anything more I should change about the PR?

**Comment by ZackerySpytz** [comment]
Please add an entry to `Doc/whatsnew/3.10.rst` and add `versionadded` directives to the documentation for the new functions.

**Comment by taleinat** [comment]
> @stevendaprano I'd be _against_ having `predict_x` as this is incorrect if this is a statistical model y = a + bx + noise, because it ignores the noise.

I'm also -1 on adding `.predict_{x,y}` methods. I feel that they would be beyond the scope of what this module sets out to achieve, and they'd also be surprising for many to find on a named tuple.

File: Doc/whatsnew/3.10.rst
Comment (taleinat) [inline Doc/whatsnew/3.10.rst:224]:
```suggestion
functions to the :mod:`statistics` module.
```

Comment (taleinat) [inline Doc/whatsnew/3.10.rst]:
```suggestion
Added :func:`~statistics.covariance`, Pearson's
:func:`~statistics.correlation`, and simple
:func:`~statistics.linear_regression` functions.
```

**Comment by taleinat** [comment]
Apart from my minor suggestion for the What's New entry, this looks good to me.

@rhettinger, would you like to take another look? Any other comments?

File: Doc/library/statistics.rst
Comment (stevendaprano) [inline Doc/library/statistics.rst]: The markup here seems to be broken.

    *dependent_variable = intercept + slope * regressor + noise*

Should be:

    *dependent_variable = intercept + slope \* regressor + noise*

Untested; if the backslash doesn't work, try verbatim mode

    ```*```

to escape the multiplication sign.

File: Misc/NEWS.d/next/Library/2019-10-16-08-08-14.bpo-38490.QbDXEF.rst
Comment (stevendaprano) [inline Misc/NEWS.d/next/Library/2019-10-16-08-08-14.bpo-38490.QbDXEF.rst:1]: Did you mean "functionality" or "functions"? I am good with either choice, but "functions" might be slightly less unusual.

**Review by stevendaprano — COMMENTED** [review]
@twolodzko I am very impressed in this, thank you, and I look forward to using it. I just commented on what looks like a broken piece of ReST and a slightly unusual choice of wording, but apart from this I am very happy with your patch and I think it should be merged.

File: Misc/NEWS.d/next/Library/2019-10-16-08-08-14.bpo-38490.QbDXEF.rst
Comment (taleinat) [reply Misc/NEWS.d/next/Library/2019-10-16-08-08-14.bpo-38490.QbDXEF.rst:1]: @twolodzko has already applied my suggested changes to that, which include replacing "functionality" with "functions".

File: Doc/library/statistics.rst
Comment (taleinat) [reply Doc/library/statistics.rst]: Actually, the existing markup works as intended, because asterisks surrounded by whitespace are apparently ignored. See [the docs](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#inline-markup) for some more details.

I still think that explicitly escaping the middle asterisk, as suggested by @stevendaprano, would be a good idea to make it's intention clear and for future-proofing.

**Comment by twolodzko** [comment]
If I'm not missing something, all issues seem to be resolved right now.

**Comment by taleinat** [comment]
> If I'm not missing something, all issues seem to be resolved right now.

Indeed.

**Comment by twolodzko** [comment]
It's been over 20 days since last discussion on this PR. I don't want to be pushy, but this PR was posted over a year ago and at some point I would forget about it as well, and all the time we all spend on it would be wasted. If there's anything I can do about it, I'm open to it, but right now it is not exactly clear for me if we are waiting for something?

**Comment by ZackerySpytz** [comment]
Unfortunately, it seems like there are some minor grammatical errors in the new docstrings and documentation.

**Comment by twolodzko** [comment]
@ZackerySpytz could you be more specific?

**Comment by twolodzko** [comment]
I don't know what is to be fixed. Hope this will get merged at some point. If there's anything I need to do, please tag me.

**Comment by twolodzko** [comment]
@taleinat is there anything blocking? Sorry for pinging you, but you seem to be the only one who cares about this PR.

**Comment by twolodzko** [comment]
This PR is ready. It waits for reviews and merge. If there is anything I can do about it, please let me know.

**Comment by taleinat** [comment]
Merged!

This has indeed reached a very good state a while ago and was reviewed favorably by several core devs. Definitely good to have this for 3.10.

**Pull request merged**
